### ISSUE TITLE
Add MacOS compilation to Wiki

### ISSUE THREAD

**Issue by AaronMT** [sm64pc/sm64pc#235]
Feel free to add a wiki page for MacOS compilation:

The game can be compiled on MacOS (10.15.4 as of this writing) with the following installed requirements:

* make
* mingw-w64
* gcc
* sdl2
* pkg-config
* glew
* glfw3
* libusb
 
```% brew install make mingw-w64 gcc sdl2 pkg-config glew glfw3 libusb```

Compile via ```gmake OSX_BUILD=1``` with ```./baserom.<region>.z64``` available for asset extraction

Run via ```./build/<region>_pc/sm64.<region>.f3dex2e```

**Comment by Zoltar007** [comment]
A package installer (Homebrew, MacPorts, etc), along with Xcode Command Line Tools are also required.

**Comment by superbonaci** [comment]
What are the available strings for <region>?

**Comment by Zoltar007** [comment]
> What are the available strings for ?
@superbonaci There's no mention of strings in this issue. I'm not sure what you're asking.

**Comment by superbonaci** [comment]
Sorry typo because of GitHub markdown. I meant for `<region>`, which are described here:
https://github.com/sm64pc/sm64pc/blob/master/extract_assets.py#L77
````
all_langs = ["jp", "us", "eu", "sh"]
````

**Comment by Zoltar007** [comment]
Oh, yes. That's exactly what `<region>` relates to.

**Comment by i82orbom** [comment]
Probably add that you need to execute the resulting binary with ./sm64.<region> -AppleMagnifiedMode YES otherwise the OS scaling will mess with the graphics.

**Comment by superbonaci** [comment]
Does gmake set +x flag automatically for out binary?

**Comment by i82orbom** [comment]
> Does gmake set +x flag automatically for out binary?

It does

**Comment by shaunlebron** [comment]
Thanks for the instructions, the compilation also required this for me:

```
brew install audiofile
```

**Comment by superbonaci** [comment]
> Thanks for the instructions, the compilation also required this for me:
> 
> ```
> brew install audiofile
> ```

Did you run `make clean` every time before make?
https://github.com/sm64pc/sm64pc/issues/26

**Comment by shaunlebron** [comment]
it compiled fine, the audiofile library was just missing from the original list of brew formulas

**Comment by superbonaci** [comment]
It's curious because I didn't need it.

**Comment by gmipf** [comment]
Have tried it on 10.13.6, runs fine.

**Comment by i82orbom** [comment]
audiofile is necesary. @superbonaci you did not need it probably because you had it already installed, or another package installed it as dependency most probably

**Comment by gmipf** [comment]
Can I compile a version with german language? have tried compiling using the eu rom but he asked for the us rom.

**Comment by coltongit** [comment]
> Can I compile a version with german language? have tried compiling using the eu rom but he asked for the us rom.

The German language is only in the EU release. You'll have to compile using the EU ROM.

**Comment by gmipf** [comment]
OK, got it now, I had to build using: 

`gmake OSX_BUILD=1 VERSION=eu`

But now something is messed up, I think because of the us savefile, where is the save located?

**Comment by i82orbom** [comment]
It's located where you executed the binary, in your case it should be in ./build/us_pc/sm64_save_file.bin

**Comment by gmipf** [comment]
No it isn't there and the eu version loads the savefile for the us executable...

**Comment by gmipf** [comment]
OK got it, it is at the root of the source directory. But still messed up, there is no sound in eu binary.

**Comment by i82orbom** [comment]
Yeah, I tried to build the eu version once and I had some problems as well, in my case it was no sound.... Maybe you could try with the US version

**Comment by coltongit** [comment]
No sound on EU is an issue we're still trying to fix atm. At least the text works now unlike in master

**Comment by gmipf** [comment]
US version runs fine, was the first version I have built.

**Comment by gmipf** [comment]
I also noticed the eu binary runs at "50Hz" speed. So there is no point using the eu version then anyway. Adding the languages to the us version is the way to go I think.

**Comment by halpz** [comment]
how can I package the build in an .app bundle?

**Comment by Zoltar007** [comment]
> how can I package the build in an .app bundle?

You'd either need to create the folders & the `info.plist` file manually, or via XCode. There are multiple different ways to do this.
Then you'd need to bundle in the dynamic libraries via `otool` or a script that automates it like [mac dylib bundler](https://github.com/auriamg/macdylibbundler/).

**Comment by halpz** [comment]
> > how can I package the build in an .app bundle?
> 
> You'd either need to create the folders & the `info.plist` file manually, or via XCode. There are multiple different ways to do this. Here is the [Apple Developer page](https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFBundles/BundleTypes/BundleTypes.html#//apple_ref/doc/uid/10000123i-CH101-SW1) about it.
> Then you'd need to bundle in the dynamic libraries via `otool` or a script that automates it like [mac dylib bundler](https://github.com/auriamg/macdylibbundler/).

I've bundled the app and the dylibs, but it just crashes on launch (Death of untracked active application: <private>)
it still runs if I launch the file directly , but not from launching the .app-
anyone had success with this?

**Comment by superbonaci** [comment]
> audiofile is necessary. @superbonaci you did not need it probably because you had it already installed, or another package installed it as dependency most probably. So it should be added to the list.

You are right, tried compiling without `audiofile` and throws error. It's mandatory to have it.

**Comment by superbonaci** [comment]
> how can I package the build in an .app bundle?

You can try [appify](https://github.com/machinebox/appify), but I'm not sure savegame will work.

**Comment by Zoltar007** [comment]
This is only an issue with the Master branch. In the Nightly branch, the savegame & config file location were changed to `~\Application Support\sm64pc`. If you're still using the Master branch, you can get around the issue with the savegame & config files failing to create inside the app bundle by creating a shell script with the following contents, then point the `CFBundleExecutable` key in the `Info.plist` file at it instead of the binary:
```
#!/bin/bash
cd "${0%/*}"
./sm64.us.f3dex2e
```

**Comment by superbonaci** [comment]
Are you talking about the Nightly branch of sm64pc or appify?

**Comment by Zoltar007** [comment]
This repository. There's no issue with appify.

**Comment by superbonaci** [comment]
Ok how can I `git clone` or `git pull` nightly branch to test?

**Comment by halpz** [comment]
I used this to make an app bundle: https://github.com/ProgramYazar/MacBundleCreator
then this to bundle dylibs: https://github.com/auriamg/macdylibbundler/
then edited the 'RunIT' script to read like this:
```
#!/bin/sh
"$(dirname "$0")"/sm64.us.f3dex2e "$@"
exit 0
```
and works perfectly as an app bundle

**Comment by Zoltar007** [comment]
> Ok how can I `git clone` or `git pull` nightly branch to test?

`git checkout nightly`

**Comment by halpz** [comment]
can you build a mac executable on a windows machine? or do you have to build on the intended platform?

**Comment by superbonaci** [comment]
> 
> 
> can you build a mac executable on a windows machine? or do you have to build on the intended platform?

It depends if you can cross compile for macOS from either Windows or Linux.

**Comment by halpz** [comment]
> > can you build a mac executable on a windows machine? or do you have to build on the intended platform?
> 
> It depends if you can cross compile for macOS from either Windows or Linux.

I guess that's what I'm asking

**Comment by superbonaci** [comment]
> 
> 
> > Ok how can I `git clone` or `git pull` nightly branch to test?
> 
> `git checkout nightly`

Tested and working with `appify` and nightly build. Now files are saved to:
````
/Users/$USER/Library/Application Support/sm64pc/sm64_save_file.bin
/Users/$USER/Library/Application Support/sm64pc/sm64config.txt
````
You can call the .app any name and move it to any location, savegame and key configuration will remain working.

**Comment by i82orbom** [comment]
> > > can you build a mac executable on a windows machine? or do you have to build on the intended platform?
> > 
> > 
> > It depends if you can cross compile for macOS from either Windows or Linux.
> 
> I guess that's what I'm asking

You can cross compile from Mac to windows or linux... technically is possible to cross compile from Windows and generate a mach-o?? output from windows... or linux, but never tried myself

**Comment by superbonaci** [comment]
How to make the app bundle:

1.Install brew
2.Install brew formulas
3.git clone, git checkout nightly, git pull
4.Compile
5.Make bundle with appify or any other app. appify requires Go

How to use appify:
````
cd /Applications
~/go/bin/appify -name "SuperMario64" -icon ~/Downloads/mario64.png ~/sm64pc/build/us_pc/sm64.us.f3dex2e 
````

**Comment by i82orbom** [comment]
I think the building and the bundling instructions should be separated, as you mentioned if you require go for appify... then you'll need an extra

`brew install go`

**Comment by superbonaci** [comment]
> 
> 
> can you build a mac executable on a windows machine? or do you have to build on the intended platform?

Did you already compile sm64pc on Windows? You can try recompiling with `gmake OSX_BUILD=1`.
If somebody here has the build environment already setup on Windows could try and help us.

**Comment by superbonaci** [comment]
> 
> 
> I think the building and the bundling instructions should be separated, as you mentioned if you require go for appify... then you'll need an extra
> 
> `brew install go`
> 
> Plus installing appify:
> 
> `go get github.com/machinebox/appify`

I did not install go with brew, just downloaded [go1.14.4.darwin-amd64.pkg](https://golang.org/dl/) installer from website. I guess it can be done in several ways.

**Comment by i82orbom** [comment]
@superbonaci True that... hopefully that package sets GOTPATH and the required env vars in the profile... not sure though if this is happening in mac.

In Linux you have to do it yourself, so when installing apps through go get 'packagename' if you don't have GOROOT in your path, you won't be able to use installed cli-tools unless you reference them using the full path, keep this in mind if this is your case.

**Comment by superbonaci** [comment]
@i82orbom it must set the vars because before installing Go I got an error and after installing it appify worked fine.

**Comment by sharpSteff** [comment]
I tried to build on macOS Catalina, build got errors when running make `OSX_BUILD=1 VERSION=us`:

`In file included from audiofile.cpp:5006:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:317:9: error: 
      no member named 'signbit' in the global namespace
using ::signbit;

177 warnings and 13 errors generated.
make[1]: *** [Makefile:7: audiofile.o] Error 1
make[1]: Leaving directory '/repos/sm64ex/tools/audiofile'
make: *** [Makefile:41: audiofile/libaudiofile.a] Error 2
make: Leaving directory '/repos/sm64ex/tools'
Makefile:242: *** Failed to build tools.  Stop.
`

any ideas how to solve this ?

**Comment by superbonaci** [comment]
@sharpSteff which macOS are you running? What does output `brew list`?

**Comment by sharpSteff** [comment]
macOS Catalina 10.15.5 

brew list                    
cairo			glfw			kf5-attica		openssl@1.1
cmake			glib			libffi			pango
dbus			gmp			libmpc			pcre
doxygen			graphite2		libpng			perl
fontconfig		graphviz		libtiff			pixman
freetype		gts			libtool			pkg-config
fribidi			harfbuzz		libusb			python@3.8
gcc			icu4c			lzo			qt
gcc@7			isl			make			readline
gd			jasper			mingw-w64		sdl2
gdbm			jpeg			mpfr			sqlite
gettext			kde-extra-cmake-modules	netpbm			webp
glew			kde-karchive		ninja			xz

**Comment by sharpSteff** [comment]
I think I got it.
I had to install gcc@9 as well, since CC is defined as "cpp-9" for macOS 

I also had to replace to default gcc in tools/audiofile with `/usr/local/bin/gcc-10`

**Comment by coltongit** [comment]
> I think I got it.
> I had to install gcc@9 as well, since CC is defined as "cpp-9" for macOS in the main MakeFile
> 
> I also had to replace to default gcc in tools/audiofile with `/usr/local/bin/gcc-10`

changing cpp-9 to cpp might help with that in your case

**Comment by superbonaci** [comment]
@sharpSteff so it's fixed for you?

**Comment by tosih** [comment]
Installing `gcc@9` fixed it for me.

**Comment by beyondmeat** [comment]
Is there anyway to have the sound and res folder inside the app bundle with external data = 1. Right now I had to copy those folders to ~/Library/Application Support/sm64ex for the app bundle to run. If I compile with external data, this isn't an issue.

**Comment by Zoltar007** [comment]
> Is there anyway to have the sound and res folder inside the app bundle with external data = 1. Right now I had to copy those folders to ~/Library/Application Support/sm64ex for the app bundle to run. If I compile with external data, this isn't an issue.

EDIT: I misread the quote. As far as I'm aware, you'll need to manually copy the sound & res folders. The below is if you want to include any dependencies for moving to another computer that doesn't have the build environment set up.

Use [mac dylib bundler](https://github.com/auriamg/macdylibbundler) to bundle libraries. Here's two examples of one line commands:

Stand-alone binary:
`dylibbundler -od -b -x ./sm64.us.f3dex2e -d ./libs -p ./libs`

In an app bundle:
`dylibbundler -od -b -x /Applications/SuperMario64PC.app/Contents/MacOS/sm64.us.f3dex2e -d /Applications/SuperMario64PC.app/Contents/libs`

You will need to include the new "libs" folder with the binary.

**Comment by superbonaci** [comment]
What about compiling on the new `Apple Silicon` hardware? Does it generate dual architecture binary for both ARM and INTEL architectures?
